HARP-14633: island geometry disappears on certain zoom level

dbacarel opened the pull request:
Currently, in master, all the ring geometries forming multi-polygons are processed and put into a single polygon composed of an interleaved list of exterior/interior rings. The winding (determined in function of the sign of the area) of these rings is used to differentiate exteriors from interiors.

This layout besides being not fully compliant to the MVT spec, see [Polygon Geometry Type](https://github.com/mapbox/vector-tile-spec/tree/master/2.1#4344-polygon-geometry-type), creates confusion in case clipped rings result in a `0` area.


This PR proposes to process and create polygon objects in accordance with the MVT spec while stopping relying on the winding.

atomicsulfate reviewed (changes requested):
Add some IBCT tests:
- In GeoJsonFeaturesRendering as I already mentioned, to catch the winding bug in GeoJsonDataAdapter.
- In mapsdk/@here/integration-tests/test/ibct/OmvMapRender.ts, add a test with the actual tile that was having the issue in China OMV. For that, download the omv tile to a file (you can get it from the developer console) and use it like the other tests in that suite do.

I think (slight) changes in extruded building IBCT are expected, since polygons are split differently. I don't see any big difference , maybe a look more carefully in the next review iteration.

dbacarel commented:
rebuild please

dbacarel commented:
retest this please

dbacarel commented on @here/harp-vectortile-datasource/lib/adapters/geojson-vt/GeoJsonVtDataAdapter.ts:
```diff
@@ -181,20 +181,29 @@ export class GeoJsonVtDataAdapter implements DataAdapter {
                 case VTJsonGeometryType.Polygon: {
-                    const polygon: IPolygonGeometry = { rings: [] };
+                    const polygons: IPolygonGeometry[] = [];
+                    let polygon: IPolygonGeometry = { rings: [] };
```
done

dbacarel commented on @here/harp-vectortile-datasource/lib/adapters/geojson-vt/GeoJsonVtDataAdapter.ts:
```diff
@@ -181,20 +181,29 @@ export class GeoJsonVtDataAdapter implements DataAdapter {
+                            polygons.push(polygon);
+                        }
+                        // Push the ring into the current polygon
                         polygon.rings.push(ring);
```
done

dbacarel commented on @here/harp-vectortile-datasource/lib/adapters/omv/OmvDataAdapter.ts:
```diff
@@ -406,7 +418,7 @@ export class OmvDataAdapter implements DataAdapter, OmvVisitor {
 
         const geometry: IPolygonGeometry[] = [];
-        const currentPolygon: IPolygonGeometry = { rings: [] };
+        let currentPolygon: IPolygonGeometry = { rings: [] };
```
done

dbacarel commented on @here/harp-vectortile-datasource/lib/adapters/omv/OmvDataAdapter.ts:
```diff
@@ -153,26 +153,38 @@ export function asGeometryType(feature: com.mapbox.pb.Tile.IFeature | undefined)
     const firstPolygon = multipolygon[0];
 
-    if (firstPolygon.rings.length === 0) {
+    if (!firstPolygon || !firstPolygon.rings.length) {
```
done

dbacarel commented on @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts:
```diff
@@ -845,10 +845,9 @@ export class VectorTileDataEmitter {
-
+                let discardPolygon: boolean = false;
+                for (let i = 0; i < polygon.rings.length; ) {
+                    let ringCoords = polygon.rings[i++];
```
it was a left-over from a previous change where I needed the ring's index. However, in the context of the simplification of the nested check (as per the comment below), I left this notation so that it's possible to derive which ring is exterior from its index.

dbacarel commented on @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts:
```diff
@@ -845,10 +845,9 @@ export class VectorTileDataEmitter {
-                for (const outline of polygon.rings) {
-                    let coords = outline;
-
+                let discardPolygon: boolean = false;
```
turns out that `discardPolygon` wasn't needed, after all. I reshuffled the logic a bit and even without moving out it to a separate function, I think it's easier to read/follow now.

dbacarel commented on @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts:
```diff
@@ -857,33 +856,57 @@ export class VectorTileDataEmitter {
+                            // can imply that all the inner rings collapsed as well.
+                            // As per spec, all the inner rings must enclosed within the exterior ring.
+                            // Using this assumption, we can skip the whole polygon.
+                            if (!THREE.ShapeUtils.area(ringCoords)) {
```
that line doesn't exist anymore, though all the rest are now explicit.

dbacarel commented on @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts:
```diff
@@ -857,33 +856,57 @@ export class VectorTileDataEmitter {
+                            if (!THREE.ShapeUtils.area(ringCoords)) {
+                                if (winding > 0) {
+                                    discardPolygon = true;
+                                    break;
```
I reshuffled the logic a bit, there is now only the `break` left and in my opinion, the flow is clearer now. Happy to change it if it's not the case

dbacarel commented on @here/harp-vectortile-datasource/lib/adapters/omv/OmvDataAdapter.ts:
```diff
@@ -417,6 +429,14 @@ export class OmvDataAdapter implements DataAdapter, OmvVisitor {
                     if (currentRing !== undefined && currentRing.length > 0) {
+                        // MVT spec defines that each exterior ring signals the beginning of a new polygon.
+                        // see https://github.com/mapbox/vector-tile-spec/tree/master/2.1
+                        if (ShapeUtils.area(currentRing) > 0) {
```
done.

dbacarel commented on @here/harp-vectortile-datasource/lib/adapters/geojson/GeoJsonDataAdapter.ts:
```diff
@@ -134,21 +134,11 @@ function convertLineGeometry(
         const isOuterRing = i === 0;
-        const isClockWise = signedPolygonArea(ring) < 0;
         const { positions } = convertLineStringGeometry(ring, decodeInfo);
+        const isClockWise = ShapeUtils.area(positions) > 0;
```
Nothing changed here actually, I moved the winding check  from before the projection to tile coordinate

```js
// ShapeUtils.area returns negative area for CW ring with Y axis increasing upwards.
const isClockWise = ShapeUtils.area(positions) < 0; 
const { positions } = convertLineStringGeometry(ring, decodeInfo);
``` 
 to after the projection

 ```js
const { positions } = convertLineStringGeometry(ring, decodeInfo);
 // ShapeUtils.area returns positive area for CW ring in tile coords - Y axis increasing downwards
const isClockWise = ShapeUtils.area(positions) > 0; 
 ```
 
So the sign of the area is of course inverted in tile coordinates as you pointed out.
There are already tests in `GeoJsonFeaturesRendering.ts` checking on polygons with wrong winding

dbacarel commented:
> Add some IBCT tests:
> 
> * In GeoJsonFeaturesRendering as I already mentioned, to catch the winding bug in GeoJsonDataAdapter.
> * In mapsdk/@here/integration-tests/test/ibct/OmvMapRender.ts, add a test with the actual tile that was having the issue in China OMV. For that, download the omv tile to a file (you can get it from the developer console) and use it like the other tests in that suite do.
> 
> I think (slight) changes in extruded building IBCT are expected, since polygons are split differently. I don't see any big difference , maybe a look more carefully in the next review iteration.


- I added an IBCT in `test/ibct/OmvMapRender.ts` to keep track of the issue in China OMV.
- I didn't add any additional IBCT test in `GeoJsonFeaturesRendering` because checks are already there, I explained my reasons up in the comment. Let me know if they are not sound.


Regarding the previous IBCT failing tests, they are now fine - it's failing now because of the newly added IBCT test not having yet a reference image. We can push it once the review is completed I guess.
 
Previously, some tests were failing because of this extra polygon being pushed in `OmvDataAdapater.ts` 
```js
        if (currentPolygon.rings.length > 0) {
            geometry.push(currentPolygon);
        }
```
I forgot to remove it when I changed the logic during the conversion from MVT

atomicsulfate commented on @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts:
```diff
@@ -845,45 +845,71 @@ export class VectorTileDataEmitter {
 
-                if (rings.length === 0) {
-                    continue;
+                if (rings.length) {
```
```suggestion
                if (rings.length > 0) {
```

dbacarel replied:
done

atomicsulfate commented on @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts:
```diff
@@ -1388,288 +1414,283 @@ export class VectorTileDataEmitter {
+                    );
                 }
+            }
 
```
wow,  the diff got totally messed up beyond this point. I assume there's no changes beyond this point (apart from indentation) is that right?

dbacarel replied:
correct, nothing to check beyond here

atomicsulfate commented on @here/harp-vectortile-datasource/lib/adapters/geojson-vt/GeoJsonVtDataAdapter.ts:
```diff
@@ -181,20 +181,29 @@ export class GeoJsonVtDataAdapter implements DataAdapter {
                 }
                 case VTJsonGeometryType.Polygon: {
-                    const polygon: IPolygonGeometry = { rings: [] };
+                    const polygons: IPolygonGeometry[] = [];
```
Could you add a test in GeoJsonVtAdapterTest to check that nested rings are separated into polygons?

dbacarel replied:
done

atomicsulfate commented on @here/harp-vectortile-datasource/lib/adapters/omv/OmvDataAdapter.ts:
```diff
@@ -417,17 +430,26 @@ export class OmvDataAdapter implements DataAdapter, OmvVisitor {
                     currentRing.push(command.position);
                 } else if (isClosePathCommand(command)) {
                     if (currentRing !== undefined && currentRing.length > 0) {
+                        const currentRingWinding = Math.sign(ShapeUtils.area(currentRing));
```
Could you add a unit test to cover this?

dbacarel replied:
done - I created `OmvDataAdapterTest.ts` testing 2 MultiPolygons: one with correct winding and another with opposite winding

dbacarel commented:
rebase please
